File: package/python/src/flood_fill.py
def flood_fill(
    image: list[list[int]], sr: int, sc: int, new_color: int
) -> list[list[int]]:
    """
    画像内の特定の領域を、新しい色で塗りつぶす。

    Args:
        image (list[list[int]]): 2次元のグリッド画像
        sr (int): 開始行のインデックス
        sc (int): 開始列のインデックス
        new_color (int): 塗りつぶしに使用する色

    Returns:
        list[list[int]]: 塗りつぶし後の画像
    """
    if not image or not image[0]:
        return image

    rows, cols = len(image), len(image[0])
    original_color = image[sr][sc]

    # If the starting color is already the new color, no change needed
    if original_color == new_color:
        return image

    # Create a deep copy of the image to avoid modifying the original
    result = [row[:] for row in image]

    def dfs(r: int, c: int) -> None:
        # Check bounds and if current cell has the original color
        if r < 0 or r >= rows or c < 0 or c >= cols or result[r][c] != original_color:
            return

        # Change the color
        result[r][c] = new_color

        # Recursively fill adjacent cells (up, down, left, right)
        dfs(r - 1, c)  # up
        dfs(r + 1, c)  # down
        dfs(r, c - 1)  # left
        dfs(r, c + 1)  # right

    dfs(sr, sc)
    return result

File: package/typescript/src/floodFill.ts
/**
 * 画像内の特定の領域を、新しい色で塗りつぶす。
 *
 * @param image - 2次元のグリッド画像
 * @param sr - 開始行のインデックス
 * @param sc - 開始列のインデックス
 * @param newColor - 塗りつぶしに使用する色
 * @returns 塗りつぶし後の画像
 */
export function floodFill(
  image: number[][],
  sr: number,
  sc: number,
  newColor: number
): number[][] {
  if (!image || !image[0]) {
    return image;
  }

  const rows = image.length;
  const cols = image[0].length;
  const originalColor = image[sr][sc];

  // If the starting color is already the new color, no change needed
  if (originalColor === newColor) {
    return image;
  }

  // Create a deep copy of the image to avoid modifying the original
  const result = image.map(row => [...row]);

  function dfs(r: number, c: number): void {
    // Check bounds and if current cell has the original color
    if (r < 0 || r >= rows || c < 0 || c >= cols || result[r][c] !== originalColor) {
      return;
    }

    // Change the color
    result[r][c] = newColor;

    // Recursively fill adjacent cells (up, down, left, right)
    dfs(r - 1, c); // up
    dfs(r + 1, c); // down
    dfs(r, c - 1); // left
    dfs(r, c + 1); // right
  }

  dfs(sr, sc);
  return result;
}
